src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs

```diff
@@ -262,10 +262,7 @@ public static void AddProviderTransparent(TypeDescriptionProvider provider, obje
         /// </summary>
         private static void CheckDefaultProvider(Type type)
         {
-            if (s_defaultProviders.ContainsKey(type))
-            {
-                return;
-            }
+            bool providerAdded = false;
```

Comment:
See my comment in Conversations - performance here will be super slow due to `GetCustomAttributes` + the `lock` coverage.

Reply:
May I ask what is the correct way to benchmark this?

| Method          | Toolchain    | Mean         | Error      | StdDev     | Ratio | RatioSD | Gen0    | Gen1   | Gen2   | Allocated | Alloc Ratio |
|---------------- |------------- |-------------:|-----------:|-----------:|------:|--------:|--------:|-------:|-------:|----------:|------------:|
| MultipleThreads | PR #85156    | 1,928.274 μs | 20.2352 μs | 17.9380 μs |  0.97 |    0.01 | 35.1563 | 7.8125 | 7.8125 |  15.77 KB |        1.51 |
| MultipleThreads | enlarge lock | 1,975.809 μs | 15.8019 μs | 14.7811 μs |  1.00 |    0.01 | 35.1563 | 7.8125 | 7.8125 |  10.42 KB |        1.00 |
| MultipleThreads | main repo    | 1,983.598 μs | 18.0052 μs | 16.8421 μs |  1.00 |    0.00 | 35.1563 | 7.8125 | 7.8125 |  10.43 KB |        1.00 |
|                 |              |              |            |            |       |         |         |        |        |           |             |
| SingleThread    | PR #85156    |     8.458 μs |  0.1428 μs |  0.1266 μs |  1.62 |    0.03 |  0.3052 | 0.0916 |      - |   2.04 KB |        1.04 |
| SingleThread    | enlarge lock |     5.749 μs |  0.1098 μs |  0.0973 μs |  1.10 |    0.02 |  0.3052 |      - |      - |   1.95 KB |        1.00 |
| SingleThread    | main repo    |     5.233 μs |  0.0264 μs |  0.0247 μs |  1.00 |    0.00 |  0.3052 |      - |      - |   1.95 KB |        1.00 |

<details>
<summary>code</summary>
<pre>
using BenchmarkDotNet.Attributes;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading.Tasks;

namespace PerfBm
{
    [MemoryDiagnoser]
    public class TypeDescriptorBenchmark
    {
        delegate void Clear();
        private static readonly Clear ClearAction;
        static TypeDescriptorBenchmark()
        {
            // Helper method is added to System.ComponentModel.TypeDescriptor to clear all hashtables.
            var mi = typeof(TypeDescriptor).GetMethod("ClearCache")!;
            ClearAction = mi.CreateDelegate<Clear>();

            ClearAction();
        }

        public static void CleanCache()
        {
            ClearAction();
        }

        [Benchmark]
        public void MultipleThreads()
        {
            CleanCache();

            int concurrentCount = Environment.ProcessorCount * 2;

            using var finished = new CountdownEvent(concurrentCount);

            var instances = new TypeWithProperty[concurrentCount];
            for (int i = 0; i < concurrentCount; i++)
            {
                instances[i] = new TypeWithProperty();
            }

            for (int i = 0; i < concurrentCount; i++)
            {
                int i2 = i;
                new Thread(() =>
                {
                    TypeDescriptor.GetProperties(instances[i2]);
                    finished.Signal();
                }).Start();
            }

            finished.Wait();
        }
        [Benchmark]
        public object SingleThread()
        {
            CleanCache();

            return TypeDescriptor.GetProperties(new TypeWithProperty());
        }

        [TypeDescriptionProvider(typeof(EmptyPropertiesTypeProvider))]
        public sealed class TypeWithProperty
        {
            public int OneProperty { get; set; }
        }
        public sealed class EmptyPropertiesTypeProvider : TypeDescriptionProvider
        {
            private sealed class EmptyPropertyListDescriptor : ICustomTypeDescriptor
            {
                public AttributeCollection GetAttributes() => AttributeCollection.Empty;

                public string? GetClassName() => null;

                public string? GetComponentName() => null;

                public TypeConverter? GetConverter() => new TypeConverter();

                public EventDescriptor? GetDefaultEvent() => null;

                public PropertyDescriptor? GetDefaultProperty() => null;

                public object? GetEditor(Type editorBaseType) => null;

                public EventDescriptorCollection GetEvents() => EventDescriptorCollection.Empty;

                public EventDescriptorCollection GetEvents(Attribute[]? attributes) => GetEvents();

                public PropertyDescriptorCollection GetProperties() => PropertyDescriptorCollection.Empty;

                public PropertyDescriptorCollection GetProperties(Attribute[]? attributes) => GetProperties();

                public object? GetPropertyOwner(PropertyDescriptor? pd) => null;
            }
            public override ICustomTypeDescriptor GetTypeDescriptor(Type objectType, object? instance)
            {
                return new EmptyPropertyListDescriptor();
            }
        }
    }
}
</pre>
</details>

Reply:
Unless I'm missing something, `type.GetCustomAttributes()` will always be called, which is slow. I'm concerned that for the same type, subsequent access will be slow. I'm also concerned, but less so, about multiple threads accessing the widened `lock` for different types.

So, the benchmarks here should not clear the cache and not do `new TypeWithProperty()` for every iteration -- just call `GetProperties()` on a previously cached object.

Reply:
> `type.GetCustomAttributes()` will always be called

https://github.com/dotnet/runtime/blob/e4c4296078ea36931550b0029c0f70f16be18ef6/src/libraries/System.ComponentModel.TypeConverter/src/System/ComponentModel/TypeDescriptor.cs#L269-L272

There's a check before `GetCustomAttributes` to see whether the default provider is already created. So I don't think `GetCustomAttributes` is called for subsequent invocations.

Reply:
Thanks I didn't see the second use of ContainsKey within the lock which was not removed with this PR.